### PR TITLE
User login - 401, refresh token

### LINKED ISSUE #154: React on 401 code from API
**Describe the bug**
When an authorized API request from the backend returns with status code 401 ("OAuth Token is invalid or expired"), the user should be either logged out or the OAuth Token should be refreshed from Auth0.

**Is it on a Production or Development Build?**
develop

**To Reproduce**
1. Open app after you have used it with a logged in user after some time (when your token is expired)
2. See error

**Screenshots**
Bildschirmfoto 2021-01-22 um 16 57 54

**Smartphone (please complete the following information):**
- Device: iPhone 11 simulator
- OS: iOS14.3
- Tree Mapper App Version 1.0.0

### PR DESCRIPTION
**Pull request by ankitecd**
Fixes #154 , #180 , #196 , #173  

Changes in this pull request:
- Added refresh token check and 401 error check
- Added migration of realm
- Disabled Multiple tree registration

### CODE REVIEW

File: app/components/MainScreen/index.js [hunk 1]
```diff
@@ -166,9 +209,14 @@ const MainScreen = ({ navigation }) => {
                 isUserLogin={isUserLogin}
                 testID={'btn_login'}
                 accessibilityLabel={'Login/Sign Up'}
-                photo={userPhoto}
+                photo={
+                  cdnUrls && cdnUrls.cache && userInfo.image
+                    ? `${cdnUrls.cache}/profile/avatar/${userInfo.image}`
```
Comment (norbertschuler) [inline app/components/MainScreen/index.js:214]: Be aware that despite my fix this does not work yet reliable for the development backend as the url given by the config API call are not ok for the development backend - for the _planet-webapp_ we directly configured the CDN URL within the .env at https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/d4928069530a5ee401b03c3958af29b6e53d78f6/.env.local.example#L6 and used that for building the image path at https://github.com/Plant-for-the-Planet-org/planet-webapp/blob/develop/src/utils/getImageURL.ts - I would prefer keeping the config API call as on production this should work correctly.

**Comment by norbertschuler** [comment]
@ankitecd do you plan to do further fixes in this branch or mark it ready for review?

**Comment by ankitecd** [comment]
> 
> 
> @ankitecd do you plan to do further fixes in this branch or mark it ready for review?

There are some fixes left.

File: app/actions/user.js [hunk 2]
```diff
@@ -1,93 +1,253 @@
...
+        resolve(true);
+      })
+      .catch((err) => {
+        console.error(`Error at /actions/user/auth0Logout, ${JSON.stringify(err)}`);
+        dbLog.error({
+          logType: LogTypes.USER,
+          message: 'Error while Logging Out',
+          logStack: JSON.stringify(err),
+        });
+        bugsnag.notify(err);
```
Comment (norbertschuler) [inline app/actions/user.js:127]: Auth0 logout on iOS shows a dialog to choose if you really want to logout of abort this operation. If the user aborts the logout, no error should be raised. (Compare #158)

**Review by norbertschuler — CHANGES_REQUESTED** [review]

**Comment by norbertschuler** [comment]
- [ ] Got a crash after successfully registering a single tree, then clicking "Next Tree" and in the next screen closing with the "X" button at the top:

Simulator Screen Shot - iPhone 11 - 2021-02-11 at 10 20 48Simulator Screen Shot - iPhone 11 - 2021-02-11 at 10 20 53Simulator Screen Shot - iPhone 11 - 2021-02-11 at 10 20 11

File: app/components/Common/InventoryList/index.js [hunk 3]
```diff
@@ -29,13 +29,14 @@ export default function InventoryList({ inventoryList, accessibilityLabel, inven
       renderItem={({ item }) => {
         let imageURL;
         let isOffSitePoint = false;
+        console.log('item.polygons[0].coordinates', item.polygons[0].coordinates.length);
```
Comment (norbertschuler) [inline app/components/Common/InventoryList/index.js:32]: Got a crash here (s. above).

**Review by norbertschuler — CHANGES_REQUESTED** [review]

File: app/components/Common/InventoryList/index.js (hunk 3 above)
Comment (norbertschuler) [reply app/components/Common/InventoryList/index.js:32]: Maybe remove the logging if not necessary?

File: app/components/ManageSpecies/index.js [hunk 4]
```diff
@@ -64,20 +54,18 @@ const ManageSpecies = ({
         if (userSpecies && userSpecies.length > 0) {
           specieListWithUnknown = [
             ...userSpecies,
-            { guid: 'abc', isUserSpecies: true, scientific_name: 'Unknown' },
+            { guid: 'unknown', isUserSpecies: true, scientific_name: 'Unknown' },
```
Comment (norbertschuler) [inline app/components/ManageSpecies/index.js:57]: We should add a translation resource to translate the scientific name `Unknown`.

File: app/components/ManageSpecies/index.js [hunk 5]
```diff
@@ -64,20 +54,18 @@ const ManageSpecies = ({
...
           ];
         } else {
           specieListWithUnknown = [
-            { guid: 'abc', isUserSpecies: true, scientific_name: 'Unknown' },
+            { guid: 'unknown', isUserSpecies: true, scientific_name: 'Unknown' },
```
Comment (norbertschuler) [inline app/components/ManageSpecies/index.js:61]: same here

File: app/components/ManageSpecies/index.js (hunk 4 above)
Comment (norbertschuler) [reply app/components/ManageSpecies/index.js:57]: But if we add a `i18next.t` here, this probably also has to be done at https://github.com/Plant-for-the-Planet-org/treemapper/blob/9da024767c66485739bd15f37e70752804271b2e/app/components/ManageSpecies/MySpecies.js#L60-L61 Or is it possible to change the code the way that we do not need to check against resource values?

**Review by norbertschuler — CHANGES_REQUESTED** [review]

**Comment by norbertschuler** [comment]
- [ ] I managed to register some trees with data not compatible with the backend resulting into an 400 error when trying to upload:

<details><summary>Error at: /action/upload, POST - /treemapper/plantLocations</summary>

```json
{
  "data": {
    "code": 400,
    "message": "Validation Failed",
    "errors": {
      "children": {
        "type": {
        },
        "deviceLocation": {
        },
        "plantDate": {
        },
        "registrationDate": {
        },
        "captureMode": {
        },
        "plantProject": {
        },
        "species": {
          "errors": [
            "This value should not be null."
          ]
        },
        "measurements": {
          "children": {
            "height": {
            },
            "width": {
            }
          }
        },
        "geometry": {
        },
        "tag": {
        }
      }
    }
  },
  "status": 400,
  "headers": {
    "content-type": "application/json",
    "alt-svc": "h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400",
    "cache-control": "max-age=0, must-revalidate, private",
    "server": "cloudflare",
    "expires": "Sat, 13 Feb 2021 09:51:28 GMT",
    "cf-request-id": "083c667cf60000d47b67b95000000001",
    "via": "1.1 vegur",
    "cf-cache-status": "DYNAMIC",
    "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
    "cf-ray": "620da6a7fba6d47b-HAM",
    "date": "Sat, 13 Feb 2021 09:51:28 GMT",
    "vary": "Accept-Encoding"
  },
  "config": {
    "url": "https://app-development.plant-for-the-planet.org/treemapper/plantLocations",
    "method": "post",
    "data": "{\"type\":\"single\",\"captureMode\":\"on-site\",\"deviceLocation\":{\"coordinates\":[9.9355,53.5496],\"type\":\"Point\"},\"geometry\":{\"type\":\"Point\",\"coordinates\":[9.935500144958496,53.549598693847656]},\"plantDate\":\"2021-02-13\",\"registrationDate\":\"2021-02-13\",\"plantProject\":null,\"measurements\":{\"height\":3,\"width\":3},\"species\":\"sspec_1RrA7xWWwDQDkcDifP\",\"tag\":\"epublica\"}",
    "headers": {
      "Accept": "application/json, text/plain, */*",
      "Content-Type": "application/json",
      "Authorization": "OAuth qMLcf5bJW-YqUODze4N2eT95LEhfuCx8",
      "x-session-id": "02c5b899-53ba-4e6e-8ebd-883236c8b44a"
    },
    "transformRequest": [
      null
    ],
    "transformResponse": [
      null
    ],
    "timeout": 0,
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
    "maxContentLength": -1,
    "maxBodyLength": -1
  },
  "request": {
    "UNSENT": 0,
    "OPENED": 1,
    "HEADERS_RECEIVED": 2,
    "LOADING": 3,
    "DONE": 4,
    "readyState": 4,
    "status": 400,
    "timeout": 0,
    "withCredentials": true,
    "upload": {
    },
    "_aborted": false,
    "_hasError": false,
    "_method": "POST",
    "_perfKey": "network_XMLHttpRequest_https://app-development.plant-for-the-planet.org/treemapper/plantLocations",
    "_response": "{\"code\":400,\"message\":\"Validation Failed\",\"errors\":{\"children\":{\"type\":{},\"deviceLocation\":{},\"plantDate\":{},\"registrationDate\":{},\"captureMode\":{},\"plantProject\":{},\"species\":{\"errors\":[\"This value should not be null.\"]},\"measurements\":{\"children\":{\"height\":{},\"width\":{}}},\"geometry\":{},\"tag\":{}}}}",
    "_url": "https://app-development.plant-for-the-planet.org/treemapper/plantLocations",
    "_timedOut": false,
    "_trackingName": "unknown",
    "_incrementalEvents": false,
    "responseHeaders": {
      "Content-Type": "application/json",
      "Alt-Svc": "h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400",
      "Cache-Control": "max-age=0, must-revalidate, private",
      "Server": "cloudflare",
      "Expires": "Sat, 13 Feb 2021 09:51:28 GMT",
      "cf-request-id": "083c667cf60000d47b67b95000000001",
      "Via": "1.1 vegur",
      "cf-cache-status": "DYNAMIC",
      "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
      "cf-ray": "620da6a7fba6d47b-HAM",
      "Date": "Sat, 13 Feb 2021 09:51:28 GMT",
      "Vary": "Accept-Encoding"
    },
    "_requestId": null,
    "_headers": {
      "accept": "application/json, text/plain, */*",
      "content-type": "application/json",
      "authorization": "OAuth qMLcf5bJW-YqUODze4N2eT95LEhfuCx8",
      "x-session-id": "02c5b899-53ba-4e6e-8ebd-883236c8b44a"
    },
    "_responseType": "",
    "_sent": true,
    "_lowerCaseResponseHeaders": {
      "content-type": "application/json",
      "alt-svc": "h3-27=\":443\"; ma=86400, h3-28=\":443\"; ma=86400, h3-29=\":443\"; ma=86400",
      "cache-control": "max-age=0, must-revalidate, private",
      "server": "cloudflare",
      "expires": "Sat, 13 Feb 2021 09:51:28 GMT",
      "cf-request-id": "083c667cf60000d47b67b95000000001",
      "via": "1.1 vegur",
      "cf-cache-status": "DYNAMIC",
      "expect-ct": "max-age=604800, report-uri=\"https://report-uri.cloudflare.com/cdn-cgi/beacon/expect-ct\"",
      "cf-ray": "620da6a7fba6d47b-HAM",
      "date": "Sat, 13 Feb 2021 09:51:28 GMT",
      "vary": "Accept-Encoding"
    },
    "_subscriptions": [

    ],
    "responseURL": "https://app-development.plant-for-the-planet.org/treemapper/plantLocations"
  }
}
```

</details>

**Comment by norbertschuler** [comment]
@ankitecd @sagararyal Let's finish the remaining bugs in the PR and merge it and do other feature like multiple trees et al in another pull request as otherwise this is getting to big.

File: app/components/ManageSpecies/index.js (hunk 4 above)
Comment (ankitecd) [reply app/components/ManageSpecies/index.js]: I used guid to check for same and added resource for Unknown, so that worked

**Comment by norbertschuler** [comment]
- [ ] I got an error `{"PERMISSION_DENIED": 1, "POSITION_UNAVAILABLE": 2, "TIMEOUT": 3, "code": 1, "message": "User denied access to location services."}` while opening the map to choose the position of a newly registered single tree on my iPhone 11 simulator with iOS 14.4 the first time after freshly installing the app - this might be a race condition not waiting for the user to grant the permission (just a wild guess)?

Simulator Screen Shot - iPhone 11 - 2021-02-15 at 10 36 59Simulator Screen Shot - iPhone 11 - 2021-02-15 at 10 34 31

**Comment by norbertschuler** [comment]
- [ ] I got a new kind of error trying to upload my registered trees to the development backend (@jmiridis said we should tests against the develop backend instead of staging):

```
[Mon Feb 15 2021 10:41:43.106]  ERROR    Error at /actions/upload/checkAndUploadImage, {}
[Mon Feb 15 2021 10:41:43.195]  ERROR    Error at: /action/upload, changeStatusAndUpload -> undefined
[Mon Feb 15 2021 10:41:43.198]  ERROR    undefined
```

**Comment by ankitecd** [comment]
> 
> 
>     * [ ]  I got a new kind of error trying to upload my registered trees to the development backend (@jmiridis said we should tests against the develop backend instead of staging):
> 
> 
> ```
> [Mon Feb 15 2021 10:41:43.106]  ERROR    Error at /actions/upload/checkAndUploadImage, {}
> [Mon Feb 15 2021 10:41:43.195]  ERROR    Error at: /action/upload, changeStatusAndUpload -> undefined
> [Mon Feb 15 2021 10:41:43.198]  ERROR    undefined
> ```

Actually there was a change in API response and the current code was not compatible but the previous response is what required so Jorgo is working on it. I will check why it is showing undefined as error, it will make it difficult to debug.

**Review by norbertschuler — APPROVED** [review]
LGFM:
- iOS permission error should still be solved
- CDN url problem for development system should be kept in mind

**Pull request merged**